### PR TITLE
release v2.8 rocm

### PR DESCRIPTION
**Pull request by VeeraRajasekhar**
# Description

Please include a brief summary of the changes, relevant motivation and context.

Fixes # (issue)

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

Please list the changes introduced in this PR:

- Change A
- Change B

# Checklist:

- [ ] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [ ] The functionality is complete
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: build_tools/jax.py
Comment (ipanfilo) [inline build_tools/jax.py:1]: copyright date

File: tests/pytorch/distributed/test_fusible_ops.py [hunk 1]
```diff
@@ -29,24 +29,28 @@
...
 _current_file = pathlib.Path(__file__).resolve()
 sys.path.append(str(_current_file.parent.parent))
-from utils import dtype_tols, make_recipe
+from utils import dtype_tols, make_recipe, quantization_tols
 
 
 # Check what quantization schemes are supported
 fp8_available, reason_for_no_fp8 = FP8GlobalStateManager.is_fp8_available()
 mxfp8_available, reason_for_no_mxfp8 = FP8GlobalStateManager.is_mxfp8_available()
+nvfp4_available, reason_for_no_nvfp4 = FP8GlobalStateManager.is_mxfp8_available()
```
Comment (ipanfilo) [inline tests/pytorch/distributed/test_fusible_ops.py]: Is it is_mxfp8_available in upstream? For ROCm it should be distinctive

File: transformer_engine/common/gemm/rocm_gemm.cu [hunk 2]
```diff
@@ -975,7 +975,7 @@ void hipblaslt_gemm(const Tensor *inputA,
         << " gelu=" << (outputPreGelu->data.dptr != nullptr)
         << " use_fp8=" << use_fp8
         << " scale_mode=" << (a_tensor ? "tensor" : a_block ? "mxfp8" : "unsupported")
-        << " alpha=" << alpha << " beta=" << beta
+        << " alpha=" << *reinterpret_cast<const float *>(alpha) << " beta=" << *reinterpret_cast<const float *>(beta)
```
Comment (ipanfilo) [inline transformer_engine/common/gemm/rocm_gemm.cu:978]: I wonder if it is useful. Logging pointer value does not give much info

File: transformer_engine/common/swizzle/swizzle.cu [hunk 3]
```diff
@@ -37,6 +40,7 @@ constexpr __device__ __host__ int NEW_SF_TILE_DIM_M_I32 = 32;
 // HIPCC does not support __host__ qualifier for variables
 // and constexpr values do not need __device__ qualifier because they are compile-time constants
 constexpr int MXFP8_BLOCK_SIZE = 32;
+constexpr int NVFP4_BLOCK_SIZE = 16;
```
Comment (ipanfilo) [inline transformer_engine/common/swizzle/swizzle.cu]: It can now be merged back with CUDA path since they removed __device__ __host__

**Review by ipanfilo — COMMENTED** [review]
Please update AMD copyright year on changed files

File: transformer_engine/jax/csrc/extensions/misc.h
Comment (ipanfilo) [inline transformer_engine/jax/csrc/extensions/misc.h:1]: AMD copyright

File: transformer_engine/pytorch/csrc/pybind.h
Comment (ipanfilo) [inline transformer_engine/pytorch/csrc/pybind.h:1]: AMD copyright

File: pyproject.toml [hunk 4]
```diff
@@ -5,7 +5,7 @@
 # See LICENSE for license information.
 
 [build-system]
-requires = ["setuptools>=61.0", "cmake>=3.21", "wheel", "pybind11[global]", "ninja", "pip", "torch>=2.1", "jax", "flax>=0.7.1"]
+requires = ["setuptools>=61.0", "cmake>=3.21", "wheel", "pybind11[global]", "ninja", "nvidia-mathdx==25.1.1", "pip", "torch>=2.1", "jax>=0.5.0", "flax>=0.7.1"]
```
Comment (ipanfilo) [inline pyproject.toml]: Remove nvidia-mathdx

File: tests/pytorch/attention/run_attention_with_cp.py [hunk 5]
```diff
@@ -310,24 +423,34 @@ def run_dpa_with_cp(
...
-
-    # compare results with and without CP
+        tensors_to_deq = [out, out_] if not fp8_bwd else tensors
+        for i, tensor in enumerate(tensors_to_deq):
+            tensors_to_deq[i] = tensor.dequantize()
+        if not fp8_bwd:
+            tensors[0], tensors[4] = tensors_to_deq
+    for tensor in tensors:
+        # For THD format, only check non-padding regions as gradients in padding regions are undefined
+        tensor_to_check = tensor[:cu_seqlens_q[-1]] if qkv_format == "thd" else tensor
```
Comment (Micky774) [inline tests/pytorch/attention/run_attention_with_cp.py:448]: I believe this may still include padding. Generally non-padding regions are `cu_seqlens_padded[i - 1] : cu_seqlens_padded[i] - pad_len[i - 1]`. The current `:cu_seqlens_q[-1]` works as a range _only if_ there is no padding until the end -- i.e. no internal or batch-wise padding. If that is an assumption for this test, then you can disregard this comment.

File: transformer_engine/common/gemm/rocm_gemm.cu (hunk 2 above)
Comment (VeeraRajasekhar) [reply transformer_engine/common/gemm/rocm_gemm.cu:978]: I think I added this a while ago after @wangye805  requested to print them.

@wangye805 Do you think we still need this to be printed in log?

File: tests/pytorch/distributed/test_fusible_ops.py (hunk 1 above)
Comment (VeeraRajasekhar) [reply tests/pytorch/distributed/test_fusible_ops.py]: Yes, https://github.com/NVIDIA/TransformerEngine/blob/40c69e751a47ec87786283e125c5eb264101270f/tests/pytorch/distributed/test_fusible_ops.py#L44

but in other file it is correct 
https://github.com/NVIDIA/TransformerEngine/blob/40c69e751a47ec87786283e125c5eb264101270f/tests/pytorch/test_fusible_ops.py#L48

I think this is a bug in upstream

Even dev from upstream still has it
https://github.com/NVIDIA/TransformerEngine/blob/7e48fa1bace10749e2eabe1da4bbe045bacc005d/tests/pytorch/distributed/test_fusible_ops.py#L43

File: transformer_engine/common/swizzle/swizzle.cu (hunk 3 above)
Comment (VeeraRajasekhar) [reply transformer_engine/common/swizzle/swizzle.cu]: Addressed

File: pyproject.toml (hunk 4 above)
Comment (VeeraRajasekhar) [reply pyproject.toml]: Addressed

File: tests/pytorch/attention/run_attention_with_cp.py (hunk 5 above)
Comment (VeeraRajasekhar) [reply tests/pytorch/attention/run_attention_with_cp.py:448]: Hi @Micky774, we are sending this https://github.com/ROCm/TransformerEngine/blob/c55b9dc902394686651228b26fd8f97517b6a807/tests/pytorch/attention/run_attention_with_cp.py#L126 to flash-attention.

So, as per your question, yes internal padding is there here but we are sending those details to flash-attn

and https://github.com/Dao-AILab/flash-attention/blob/a8780f2a17099fc1a3e7b00d7f5d9e08c5b71142/flash_attn/flash_attn_interface.py#L973

initially they are initialized with nan's cause of this empty_like

but if you see the https://github.com/Dao-AILab/flash-attention/blob/a8780f2a17099fc1a3e7b00d7f5d9e08c5b71142/flash_attn/flash_attn_interface.py#L337

this writes gradients to all the positions which includes internal padded positions, yes they may be garbage values still but not nan's as flash attention processed them.

as for the tail padding which i removed they were never processed by flash-attn so they may contain Nan's


@wangye805 please comment if my understanding is correct.

Comment (Micky774) [reply tests/pytorch/attention/run_attention_with_cp.py:448]: Does flash attention explicitly write to padded regions, or does it maintain them as-is? If it is the latter, then are the outputs uninitialized too? If so, then the padded regions are still a risk.

If, instead, flash attention actually _writes_ to the padded regions, what does it write? Is it the output of computation that takes junk values as input? If so, that's still a risk.

I'm just not clear on why we can confidently conclude that those values can't carry nans.

Comment (Micky774) [reply tests/pytorch/attention/run_attention_with_cp.py:448]: I chatted w/ Ye and he pointed out that the test specifically does not use internal padding for the Flash Attention backend, so this is resolved.

File: pyproject.toml
Comment (ipanfilo) [inline pyproject.toml]: Copyright date

File: tests/pytorch/distributed/test_fusible_ops.py (hunk 1 above)
Comment (ipanfilo) [reply tests/pytorch/distributed/test_fusible_ops.py]: Well, I think we should fix for us because mxfp8 and mxfp4 is supported on different HW

**Review by ipanfilo — CHANGES_REQUESTED** [review]

File: pyproject.toml
Comment (VeeraRajasekhar) [reply pyproject.toml]: Updated

File: transformer_engine/pytorch/csrc/pybind.h
Comment (VeeraRajasekhar) [reply transformer_engine/pytorch/csrc/pybind.h:1]: Updated

File: transformer_engine/jax/csrc/extensions/misc.h
Comment (VeeraRajasekhar) [reply transformer_engine/jax/csrc/extensions/misc.h:1]: Updated

**Comment by Micky774** [comment]
Looks like there's a failing GEMM test:
`OperatorTest/GEMMTestSuite.Testfp8xfp8xbf16xbf16xfp16/768x3072x4096x0x0xNNxM  # GetParam() = ((768, 3072, 4096), false, false, (false, false), 1) (Failed)`

File: transformer_engine/pytorch/fp8.py
Comment (ipanfilo) [inline transformer_engine/pytorch/fp8.py:1]: Copyright

File: transformer_engine/pytorch/fp8.py [hunk 6]
```diff
@@ -67,6 +68,18 @@ def check_mxfp8_support() -> Tuple[bool, str]:
     return False, "Device compute capability 10.0 or higher required for MXFP8 execution."
 
 
+def check_nvfp4_support() -> Tuple[bool, str]:
+    """Return if nvfp4 support is available"""
+    if IS_HIP_EXTENSION:
+        gpu_arch = get_device_compute_capability()
+        if gpu_arch >= (12, 5):  # gfx1250 and above
```
Comment (ipanfilo) [inline transformer_engine/pytorch/fp8.py]: It is not yet supported in this TE version so False should be always returned on ROCm

**Review by ipanfilo — CHANGES_REQUESTED** [review]

File: transformer_engine/pytorch/fp8.py
Comment (VeeraRajasekhar) [reply transformer_engine/pytorch/fp8.py:1]: Updated

File: transformer_engine/pytorch/fp8.py (hunk 6 above)
Comment (VeeraRajasekhar) [reply transformer_engine/pytorch/fp8.py]: Updated

**Review by ipanfilo — APPROVED** [review]

**Comment by VeeraRajasekhar** [comment]
No open review comments, I am doing an interactive rebase and cherry-picking some updates from dev.

**Comment by VeeraRajasekhar** [comment]
level-3 test run https://github.com/ROCm/TransformerEngine/actions/runs/22659347141/job/65675826930

**Comment by wenchenvincent** [comment]
@VeeraRajasekhar Why the last CI run was cancelled?
